File: notebooks/09-palindrome_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        # 1. 一路从前往后读，把所有值存进数组
        vals = []
        curr = head
        while curr:
            vals.append(curr.val)
            curr = curr.next
            
        # 2. 此时 vals 已经是像 [1, 2, 2, 1] 这样的数组了
        # 我们可以直接用双指针，从“两头向中间”靠拢比对
        left = 0
        right = len(vals) - 1
        
        while left < right:
            if vals[left] != vals[right]:
                return False  # 一旦发现不对称，立刻返回 False
            left += 1         # 左指针往右走
            right -= 1        # 右指针往左走
            
        return True

def create_linked_list(arr):
    """辅助函数：用数组快速生成链表"""
    if not arr: return None
    head = ListNode(arr[0])
    curr = head
    for val in arr[1:]:
        curr.next = ListNode(val)
        curr = curr.next
    return head

if __name__ == "__main__":
    solver = Solution()
    
    # 测试用例 1: [1, 2, 2, 1] 是回文
    head1 = create_linked_list([1, 2, 2, 1])
    print(f"测试用例 1 结果: {solver.isPalindrome(head1)}") # 应该输出 True
    
    # 测试用例 2: [1, 2] 不是回文
    head2 = create_linked_list([1, 2])
    print(f"测试用例 2 结果: {solver.isPalindrome(head2)}") # 应该输出 False

File: notebooksc++/09-palindrome_linked_list.cpp
#include <iostream>
#include <vector>
#include <cassert>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int val = 0, ListNode* next = nullptr) : val(val), next(next) {}
};

bool isPalindrome(ListNode* head) {
    vector<int> vals;
    while (head != nullptr) {
        vals.push_back(head->val);
        head = head->next;
    }
    int left = 0;
    int right = static_cast<int>(vals.size()) - 1;
    while (left < right) {
        if (vals[left] != vals[right]) {
            return false;
        }
        left++;
        right--;
    }
    return true;
}

ListNode* createLinkedList(const vector<int>& values) {
    if (values.empty()) {
        return nullptr;
    }
    ListNode* head = new ListNode(values[0]);
    ListNode* curr = head;
    for (size_t i = 1; i < values.size(); i++) {
        curr->next = new ListNode(values[i]);
        curr = curr->next;
    }
    return head;
}

void deleteLinkedList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

int main() {
    assert(isPalindrome(nullptr) == true);

    ListNode* head1 = createLinkedList({7});
    assert(isPalindrome(head1) == true);
    deleteLinkedList(head1);

    ListNode* head2 = createLinkedList({1, 2, 1});
    assert(isPalindrome(head2) == true);
    deleteLinkedList(head2);

    ListNode* head3 = createLinkedList({1, 2, 2, 1});
    assert(isPalindrome(head3) == true);
    deleteLinkedList(head3);

    ListNode* head4 = createLinkedList({1, 2, 3, 1});
    assert(isPalindrome(head4) == false);
    deleteLinkedList(head4);

    cout << "All test cases passed ✅" << endl;
    return 0;
}
